core/src/main/java/ai/timefold/solver/core/impl/domain/variable/ShadowVariableUpdateHelper.java

```diff
@@ -0,0 +1,335 @@
+import ai.timefold.solver.core.impl.domain.variable.cascade.CascadingUpdateShadowVariableDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.custom.CustomShadowVariableDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.declarative.ChangedVariableNotifier;
+import ai.timefold.solver.core.impl.domain.variable.declarative.DefaultShadowVariableSessionFactory;
+import ai.timefold.solver.core.impl.domain.variable.declarative.DefaultTopologicalOrderGraph;
+import ai.timefold.solver.core.impl.domain.variable.declarative.VariableReferenceGraph;
+import ai.timefold.solver.core.impl.domain.variable.descriptor.BasicVariableDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.index.IndexShadowVariableDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.inverserelation.InverseRelationShadowVariableDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.nextprev.NextElementShadowVariableDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.nextprev.PreviousElementShadowVariableDescriptor;
+import ai.timefold.solver.core.impl.util.Pair;
+import ai.timefold.solver.core.preview.api.domain.metamodel.ShadowVariableMetaModel;
+
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * Utility class for updating shadow variables at entity level.
+ */
+public final class ShadowVariableUpdateHelper {
+
+    private ShadowVariableUpdateHelper() {
+    }
+
+    public static <Solution_> void updateShadowVariables(Class<Solution_> solutionClass, Object... entities) {
+        var entityClassList = Arrays.stream(entities).map(Object::getClass)
+                .filter(clazz -> clazz.isAnnotationPresent(PlanningEntity.class))
+                .distinct().toList();
+        var solutionDescriptor =
+                SolutionDescriptor.buildSolutionDescriptor(Set.of(PreviewFeature.DECLARATIVE_SHADOW_VARIABLES), solutionClass,
+                        entityClassList.toArray(new Class<?>[0]));
+        var customShadowVariableDescriptorList = solutionDescriptor.getAllShadowVariableDescriptors().stream()
+                .filter(CustomShadowVariableDescriptor.class::isInstance)
+                .toList();
+        if (!customShadowVariableDescriptorList.isEmpty()) {
+            throw new IllegalArgumentException(
+                    "Custom shadow variable descriptors are not supported (%s)".formatted(customShadowVariableDescriptorList));
+        }
+        var session = new InternalShadowVariableSession<>(solutionDescriptor,
+                new VariableReferenceGraph<>(ChangedVariableNotifier.empty()));
+        session.init(entities);
+        // Update all built-in shadow variables
+        session.processBasicVariable(entities);
+        session.processChainedVariable(entities);
+        session.processListVariable(entities);
+        session.processCascadingVariable(entities);
+        session.triggerUpdateShadowVariables();
```

Comment:
I am concerned about this. This code requires a list of all shadow variables and their types. This means that - when we introduce a new type - this code will silently fail unless we remember to update it as well. 

This is brittle, and it leads me to think... why does the solver currently do all of this work without having to have all of this information in a single place? Surely, this has to be code duplication - the solver can already do this, and we just need to hook into the code which already does it now.

Reply:
The logic is primarily located at `VariableListenerSupport`. We could reuse it by creating a solution instance from the problem entity and facts. Let me check that.

Reply:
Or we could _adapt it_ so that only a part of it requires a solution.
I imagine that the full VLS gets all the variables, and the simplified one filters out those impossible without a solution.

Reply:
IMO creating new instance of a solution is a dead end. You don't know enough about the solution to be able to do that. We don't enforce public no-arg constructors, we don't enforce mutability... This is just trading one complexity for another.